File: Codes/3148-maximum-difference-score-in-a-grid.cpp
/*
 * @Author: LetMeFly
 * @Date: 2024-08-15 23:00:56
 * @LastEditors: LetMeFly
 * @LastEditTime: 2024-08-15 23:55:59
 * @Description: 这是开会的时候启的每日一题，所以看起来写了很久
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

class Solution {
public:
    int maxScore(vector<vector<int>>& grid) {
        int ans = grid[0][1] - grid[0][0];
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                int original = grid[i][j];
                if (i > 0) {
                    grid[i][j] = min(grid[i][j], grid[i - 1][j]);
                    ans = max(ans, original - grid[i - 1][j]);
                }
                if (j > 0) {
                    grid[i][j] = min(grid[i][j], grid[i][j - 1]);
                    ans = max(ans, original - grid[i][j - 1]);
                }
            }
        }
        return ans;
    }
};

File: Codes/3148-maximum-difference-score-in-a-grid.py
'''
Author: LetMeFly
Date: 2024-08-15 23:57:36
LastEditors: LetMeFly
LastEditTime: 2024-08-15 23:59:16
'''
from typing import List

class Solution:
    def maxScore(self, grid: List[List[int]]) -> int:
        ans = grid[0][1] - grid[0][0]
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                original = grid[i][j]
                if i > 0:
                    grid[i][j] = min(grid[i][j], grid[i - 1][j])
                    ans = max(ans, original - grid[i - 1][j])
                if j > 0:
                    grid[i][j] = min(grid[i][j], grid[i][j - 1])
                    ans = max(ans, original - grid[i][j - 1])
        return ans